### ISSUE TITLE
C code and luaL_error versus showError

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#477]
#473 got me thinking...

Let me just start by saying up front, I know I am guilty of this... maybe I'm the only one, but if not, I put this forward as a suggested best practice.  Unless the general consensus is strongly otherwise, I plan to change my habits and adjust it as I notice it in other code...

When C code encounters an error -- say the wrong variable type is passed, or a required argument is missing, etc., then I believe we should be terminating the function/method with `return luaL_error(L, ...)`  This terminates the function/method and allows for a formatted error string with a variable number of arguments in the error.

When a lua callback is invoked from something that C code created (timer, eventtap, etc.), it should return errors via the showError function as defined in Hammerspoon.h.  But _only_ callbacks should handle errors this way.

Any other method/function is invoked within the context of other lua or c code, and it is up to that code to determine if it should wrap its calls in `pcall` or `xpcall` (if lua) or lua_pcall (if c) to catch and handle errors, or let the error terminate execution of whatever is going on.

As I said, I know I'm guilty of sticking showError in in places it doesn't really belong, but I plan to stop.

Thoughts?  Objections?

**Comment by cmsj** [comment]
I am sure I am guilty of the same and worse - sometimes in a rush I've just made C code call NSLog and return!

This seems to me like a prime candidate for objc methods in LuaSkin. We could have something like functionError and callbackError that take nice NSString formats&values, and functionErrorWithCString and callbackErrorWithCString that take C string formats&values.

Internally they can do all the things we want - maybe they log to Crashlytics, etc.

**Comment by cmsj** [comment]
Oh, and I'd note that LuaSkin already has a method that does a pcall with debug.traceback as its error handler, which we could extend to have error handling options.

**Comment by cmsj** [comment]
tl;dr, LuaSkin exists to provide convenient wrappers to our common lua API patterns :)

**Comment by cmsj** [comment]
This might be complicated by the fact that sometimes our methods make specific promises about their ability to return things - i.e. we sometimes still return self even in an error (I think, I haven't checked every return point).

I suppose that's just another thing for LuaSkin to think about for its API in this area - does it want to do something that terminates the current execution, or not.

**Comment by asmagill** [comment]
There is certainly a case to be made for certain methods having a default fallback if their arguments make no sense or are absent... I agree that it's case by case, but anything that could be "fixed" by an attentive program will need to capture the error and probably doesn't want the visual display that showError causes...

Right now, i don't think any of our code that looks for fields in a table complains about bogus field names or missing ones... they just silently do nothing or fall back on defaults.  I'm not sure if that's good as a rule, but makes sense for things like color, so... YMMV I guess!

At any rate, I know I overuse showError, so it's good to be aware and have some guidelines for new additions.

**Comment by lowne** [comment]
my 2c on a tiny subset of this topic:
- whenever an error can be *in any way* traced back to user error in their `init.lua` (this of course includes wrong parameter types etc.), notify and terminate - they need to fix their config (the `level` parameter cannot be fully trusted due to internally nested calls in a module's code, which is why I wanted the full traceback - that way the user can hopefully find the faulty line in `init.lua` displayed somewhere in there). Trying to be helpful and guess user intensions can be counterproductive imo when it's *outside the documented usage* - if we provide helpful shortcuts/defaults/etc., they must be documented.
- whenever the error is due to unfulfilled promises from the OS, print it out, handle it gracefully, and move on - the user should be able to keep doing their work without even looking at the console (for this latter case, I use log.e or log.w from hs.logger).

**Comment by asmagill** [comment]
While I agree in principle with these ideals, in most cases, graceful error handling is the job of the lua code, not the C-code. Take the code for capturing a rectangles or text objects color in hs,drawing... it can't tell the difference between :setTextColor(nil), :setTextColor(a) where I really meant to type A, and :setTextColor(a) where a is generated somewhere and in this case (for whatever reason) actually is nil...  But at the lua level, if I do  `result, error = pcall(blah:setTextColor(a))` then if result is nil, error could (should) contain a descriptive message I can parse or have my own handler to decide what to do with it... and maybe it does just pass it along to showError, pop up the dialog and quit... or maybe in this case, as a developer, I made a typo or maybe nil is a valid result of whatever calculation I did before but rather than check it before calling setTextColor, the code seemed easier to read by chaining it and wrapping the whole thing in pcall, relying on the error handler to prevent anything crazy from happening.

(Note, I know setTextColor is a bad example -- I actually modified it recently up so that nil or an empty table falls back to the default color of black... I did this to make possible things like {red=1} rather than requiring {red=1,blue=0,green=0,alpha=1}, but I've been wondering if maybe I made it too lax -- in any case, the point is the same:  C-code can't be a good judge of how an error should be handled -- too many variables... that's what good lua code *should* do because the developer (hopefully) knows what should be recoverable and what shouldn't be.)

Callbacks are special in that they have no lua code surrounding them when they are invoked... if they fail, their only recourse is to crash (been there too many times!) or for hammerspoon itself to do something visible and attention getting *when it happens* because sometimes timing is the only clue as to which callback might actually have the problem.  This doesn't negate the need for good detection of bad states within the C code -- the hs.hotkey repeat key timer for example will refuse to fire a new timer if an existing one is already in place, which can occur if the previous keys repeatfn callback fails for some reason (which reminds me, I should do something similar with hs.timer because right now it will repeatedly pop up the dialog for showError until the timer is manually stopped.)

Hope that clarifies my intent and where I'm coming from.

**Comment by lowne** [comment]
@asmagill my quip about graceful handling was about errors *caused by the OS layer*. 
To simplify, here's a breakdown of my (admittedly utopian) ideals - I'll use your `setTextColor` example:
- if our API promises something, but for whatever reason it can't fulfil it, that's where HS should "gracefully handle" - which means, *don't grab attention* and *don't terminate* (but still letting them know something went wrong, in the console log) - e.g. if the underlying NSView craps out at an attempted `:setTextColor`. The user trying to :setTextColor may or may not notice something isn't working; if they do, they'll open up the console and be informed of any issues (and ideally if there's anything they can do to help HS work as expected; but **they did nothing wrong** so far)
  - if we expect that to happen a lot, and know of a way to work around the problem (which could be as simple as just retrying N times), but it gets too complicated in C, then we should wrap the whole thing in Lua and do the complicated stuff there, where it's easy
  - and/or if we can distinguish between N possible causes for the problem, and want to provide a helpful and information-rich error message, but, again, it's too complicated in C, we should (again) wrap'n roll.
- if one of the aforementioned "broken promises" happens in a potentially distant *event-initiated callback* (as opposed to still event-driven but user-initiated, e.g. keyboard bindings), then there's a case we should *grab attention*, but still *not terminate*; when the user comes back to the machine they have a chance to notice something went wrong
  - speaking of, for these server-like scenarios (e.g. pathwatcher, httpserver...), there should be a way (if there isn't already, didn't check) to persist the full log for after-the-fact perusal
- any user error (that is, an error in **code that was typed by the user**), whether in the main script, a user-initiated callback or an event-initiated callback, should *grab attention* and *terminate*, because the **script doesn't follow the documented API, and it needs fixing**. *Why* exactly it doesn't follow the API is not HS's business; it's the user's. To get back to your example, if you-the-user pass `a` to `:setTextColor` after complicated calculations, and 99 out of 100 times `a` is fine, but the one time is `nil`, and `:setTextColor` doesn't like `nil`, then *you* should fix your script by checking for `nil`, even if that means losing the elegance of method chaining; you can still decide instead to wrap everything in a pcall, but you'll still get the notification (more on this below)
  - if you-the-user find that it'd be really useful for your script if :setTextColor accepted `nil`, resulting in a no-op, then you open an issue: ":setTextColor should accept nil for this and that reason", or you don the contributor hat and create a PR to that effect; such an API change *must* be reflected in the documentation for :setTextColor.
  - if changing :setTextColor to **formally** accept tables that may or may not have the right fields, *and* possibly `nil`, *and* print out useful errors otherwise, is too cumbersome in C - well, I agree, it is; so (again) you wrap the whole thing Lua-side, do the parsing and erroring there, and feed whatever is the single correctly-crafted construct to the C side (example 1, see `hs.grid.ui`; example 2, we could (and should) wrap the various :setFrame methods through #471's constructor)
  - pending said PR, if you (think) you know what you're doing, you'll be quite annoyed by the console popping out every once in a while (whenever `a` happens to be `nil`) even though you dutifully pcalled it to *avoid* such interruptions. There's a case then to add a `hs.quietErrors = true --don't notify, don't pop the console` variable to HS's API - for "elite hackers/wild experimentation only"

That last quip in double quotes above is really my rationale when this boils down to 
> do *not* annoy users (with notification and termination) by default when they did something that isn't 100% kosher, they likely know what they're doing, and if they don't, and never get a chance to find out things aren't working, well too bad for them.

vs
> *do* annoy users by default, they likely made a mistake and would like to know about it so they can learn how to fix it. 

Unless, that is, we change the documentation to something along the lines of "you can pass more or less whatever you want for this argument, HS will do its best to guess your intentions, but you better watch carefully what actually happens on your machine because you never know when it might guess wrong" :) (final :setTextColor example: should `nil` mean "I'm in a hurry" (red=0,green=0,blue=0,alpha=1)? Or should it mean "make it go away" (alpha=0)? Or should it mean "whopsie" as in let's be forgiving, and no change? Or "whopsie" as in "check your script because there's an error somewhere upstream"? Any one of these is perfectly fine; but any one of these save the last one should be properly documented as "official" API)

**Comment by asmagill** [comment]
While I think these distinctions are important so we can strive to provide a consistent experience across Hammerspoon, I think this has gotten sidetracked a bit from this threads intention... (regarding the setColor function, I'll open another thread because I do have some thoughts and questions myself about what it *should* do)

Let me try again...

Lua treats errors in the same way some programming language handle exceptions -- if the code generates an error, it throws an exception which propagates up the chain of callers until something handles it -- if nothing else does, then the ultimate main lua thread in hammerspoon does, and we get an error on the console and whatever was executing ends.

pcall is a way of "catching" that error before the main thread does... what the developer does with it (correct it and try again, log a message and go to the next line of code, display its own notification and change logic to do something else, etc.) is up to the developer.

I have no problem with the main lua thread throwing up a dialog/notification via showError() to bring the users attention to it if nothing else handles the error before hand.  I have no problem with code which catches the error before the main thread does and decides to display it via showError itself...

What I disagree with is when a low level function of a module, meant to do something simple and then return (and this applies to C or lua as the coding language of choice) decides its going display a dialog and/or completely halt execution when I could do something else with it along the chain before it hits the main thread.  If I pass in a table as the primary parameter to a function in the string module of lua, it generates an error via the same mechanism that error() does... and if nothing handles it along the way back up the caller chain, it is handled by the main lua thread.  It doesn't pop up a dialog itself.

The exception to this is a callback function because the callback function's caller chain leads not into Hammerspoon but deeper into the OS which knows nothing about lua.  We *should* provide hs.showError as the message handler to lua_call -- it's almost the last point at which Hammerspoon can act on the error at all without it being lost completely.  Anything else you want the error to cause (writing to a log file/console, etc.) would have to be handled within the callback itself -- in this context the C code is the equivalent to the lua main thread and is the last handler for anything not yet handled.

This is why I think error() in lua and luaL_error in C should not be wrapped -- they are the mechanism by which errors are "thrown back" to their callers.  This is why I think the C wrapper to a lua callback function *should* use hs.showError and every other C function which is to be a function called by other lua code should *not* use hs.showError and instead use luaL_error.  luaL_error can be caught like an exception.  hs.showError throws up a dialog regardless of whether it also sends stuff back up the call chain or not.

To use the bad example I chose before -- whatever errors setColor _should_ generate, they should *not* in and of themselves cause visual feedback *from the c-code level*.  That is for whatever called the setColor function to decide (or if not decide, then pass on to its caller).  If what called the setColor function ultimately turns out to have been a callback function, then the hs.showError message handler of the callback will create the visual feedback because nothing beyond it can or will. 

Hope that's clearer.

**Comment by cmsj** [comment]
@asmagill we might get this a bit more focussed if we more precisely define "When C code encounters an error". I think you mean that when C code is aware that it's been fed garbage from Lua (e.g. the examples you cited around incorrect argument types), not that any error has occurred anywhere in C code, right?

This is a timely discussion to be having, because one of the next things I want to tackle with LuaSkin is validating argument types in functions/methods. I'm currently thinking that you'd feed LuaSkin a very simple description of the argument types you're expecting, and it validates them, fetches them out of the lua stack and returns an array of the C values. So if that LuaSkin feature will be one of the big consumers of the ideas being discussed here :)

**Comment by lowne** [comment]
Ok, so the issue at hand is about handling errors (regardless of their nature) in a structured-exception-handling way, which I agree with.
After a bit of digging, it seems the way to achieve that is to:

1. as per OP (and ensuing multiple clarifications :) ) substitute all calls to `hs.showError` with plain `error`, for both Lua and C
2. revert hs.showError to something like this (notice the pseudo-global at the end):
```lua
  hs._notify("Hammerspoon config error") -- undecided on this line
  print(err) -- the xpcall handler is debug.traceback, which takes care of the traceback
  hs.focus()
  hs.openConsole()
  hs.__TERMINATED=true 
```
3. switch callback code in C from lua_pcall to LuaSkin's `protectedCallAndTraceback` where it wasn't already
4. add a `terminate` parameter to `protectedCallAndTraceback` (or a separate function)
5. add a check at the top of `protectedCallAndTraceback` - if hs.__TERMINATED, do nothing (which is, the script encountered a fatal error at some point in the past, therefore we stop processing all events)
6. for callbacks of the user-initiated kind (hotkey,...), call `protectedCallAndTraceback` with `terminate=true`; the function itself should set hs.__TERMINATED if things go wrong (or just call in turn `showError`)
7. for other callbacks, pass false for terminate; then `protectedCallAndTraceback` will just print the error+traceback and move along

This way, we have sane defaults for error handling, but the user can still (if I'm not mistaken) wrap their callbacks with `pcall` when they so desire and handle things themselves.

**Comment by lowne** [comment]
(Also, I apologise for derailing - more than once! - the conversation)

**Comment by lowne** [comment]
1. I still think that "A fatal error provides excellent motivation for addressing problems" (from [this](https://tools.ietf.org/html/draft-thomson-postel-was-wrong-00)), but I concede about not terminating event processing on errors, whether they're user-derived (i.e. Lua script with its callbacks) or not (e.g. "Unable to write image file: Can't wrap internal representation"). A cautious user could (I think) override `hs.showError` by grabbing the stacktrace, saving it to file and calling `os.exit()` - no more events :)

2. wrt `protectedCallAndTraceback`, isn't the point of the traceback to be displayed anyway? So unless I'm missing something, something like this
  ```C
int protectedCallbackAndError(lua_State *L, int nargs, int nresults) {
    if (lua_pcall(L, nargs, nresults, 0) != LUA_OK) {
        const char *errorMsg = lua_tostring(L,-1);
        showError(L,(char*)errorMsg);
        return NO;
    }
    return YES;
}
  ```
  should suffice (instead of always having pairs of `protectedCallAndTraceback`/`showError` in the module code); there should be room still for "custom" handling/additional error messages after the fact where necessary, e.g. (from `hotkey/internal.m`):
  ```C
if (!protectedCallbackAndError(L,0,0)) {
    // For the sake of safety, we'll invalidate any repeat timer that's running, so we don't ruin the user's day by spamming them with errors
    [keyRepeatManager stopTimer];
    showError(L,'just FYI, due to that error the repeat timer was stopped')
    return noErr;
}
  ```
  but I have no idea if this needs stack juggling to deal with `nresults>0` (and yes, that would result in *two* stacktraces (see point 4) but I don't think that'd be a problem)

3. the (rare) C code that wants to deal with callback errors itself can just use `lua_pcall`

4. the stuff above (if it makes sense at all) assumes these trivial changes in `setup.lua`:
  - `hs.showError` becomes the error handler for the xpcall and
  - `hs.showError` prints `debug.stacktrace(err,2)` itself

5. for super bonus points:
  - `hs._logerror` would print to console just like hs._logmessage, but in dark red or something like that (I got lost in `^blk` and other weird stuff when I tried to understand how that works)
  - `hs.showError` would use `hs._logerror` instead of plain print
  - color of plain print would be changed to dark grey

Anyone interested in point 5? ;)

**Comment by asmagill** [comment]
Initial thoughts:

2) I think I'd rather add a method protectedCallBackWithDisplay: then force all uses of this method to display the error... I agree it's usually the case that it will be displayed, but maybe i want to stick it into the log files via NSLog or CLS_LOG instead, or maybe re-parse the error in some way... forced user feedback should always be optional in my opinion.

3) yes, but the whole point of the method was to simplify adding the traceback for whatever use you may have for it (see 2)

5) I've actually been considering changing the entire coloring of the console to something the user can modify... a way to add a "dark mode" like the OS does in 10.10 and later.  Currently all output to the console is done via #define, so these would have to be changed to variables or even NSUserDefaults, but that shouldn't be too difficult.

At it's most basic, a ^blk {} is a function that is being passed around like an argument... there are some subtle differences between this and passing a function pointer especially wrt to what "up-values" or external references are valid and how ARC collects them, but at a high level, the description works.

**Comment by asmagill** [comment]
(I should say all color output to the console...)

**Comment by lowne** [comment]
Fair enough; I was thinking about "global" redirects when necessary *by the user* (by overriding `hs.showError`) but yes, it makes sense to have flexible infrastructure in place to let single modules handle all that, should that turn out to be necessary/useful.

**Comment by lowne** [comment]
Hmm, I guess I just don't like how we're passing the error message *and* the stacktrace to `hs.showError`, because in the "handler override" scenario it doesn't seem quite right; I think the expected thing for handlers is to get passed just the error message, if they need more complex processing than just checking that for known errors they would usually use the debug library to traverse the stack call (as opposed to string-parsing the stacktrace).
It's true that the "solution" is as simple as mentioning this in the docstring, but even then... I should make it clear (about points 2 and 3 above) that I was *not* proposing to eliminate protectedCallbackAndTraceback as it is now; but merely to implement `protectedCallbackAndError` (or `protectedCallbackWithDisplay` or whatever) and change all the `protectedCallbackAndTraceback+showError` pairs to the new all-in-one function. I'm very much in agreement with leaving `protectedCallbackAndTraceback` as-is, for present and future modules that might need the full traceback string for sending it off somewhere remote.

**Comment by asmagill** [comment]
So... howsabout...

protectedCallbackAndTraceback - as is
protectedCallbackAndError - just the error message -- no traceback (and yea, lua_pcall is almost the same thing, but this keeps things similar)
protectedCallbackAndDisplayError - just the error message, but also calls hs.showerror and then hs.showerror does the traceback

Does this miss anything?

**Comment by lowne** [comment]
:+1: ❗

**Comment by lowne** [comment]
I un-concede 😀 about terminating event processing; for context, this is what I had written:
> I still think that "A fatal error provides excellent motivation for addressing problems" (from this), but I concede about not terminating event processing on errors, whether they're user-derived (i.e. Lua script with its callbacks) or not (e.g. "Unable to write image file: Can't wrap internal representation"). A cautious user could (I think) override hs.showError by grabbing the stacktrace, saving it to file and calling os.exit() - no more events :)

In my current use case, I'm trying to diagnose windowlayouts, which means a lot of windowfilters which means *tons* of debug logging in the console window. Whenever an error occurs, by the time I get to the console window (which might involve switching Space, firing an additional cascade of callbacks) I need to carefully scroll back for five minutes to find the error message. The workaround I had described above has one important drawback: I can't live-inspect the relevant data structures *at the point of failure*.

So, in an ideal world, whenever an error occurs, everything would stop, and I could take my time to get to the console and inspect whatever went wrong with untainted data (this point is the reason why simply overriding showError to e.g. stop all the windowfilters isn't optimal, as that would destroy all the data).

To restate how this could work; given the 3 LuaSkin functions previously discussed:
- `protectedCallbackAndTraceback`
- `protectedCallbackAndError`
- `protectedCallbackAndDisplayError` (this one would be used most of the time)

each of them first checks for `hs._TERMINATED`; if set, they do *not* lua_pcall, and just return success - i.e. the event is ignored and no callback occurs.
Then we'd have an `hs.stopOnError` variable, false by default. If true, `hs.showError()` would simply set `hs._TERMINATED` after doing its thing (it seems correct that `protectedCallbackAndTraceback` and `protectedCallbackAndError` can error out with "impunity" (i.e. showError() isn't called and _TERMINATED isn't set) given their specialised uses).

With all that, I could have an error happen, check the traceback, inspect the live data etc., and then briefly "resume" events (which could further help in trying to replicate the issue etc.) by issuing `hs._TERMINATED=nil`.

But to complicate things, it wouldn't be unreasonable to have `hs.hotkey` (and maybe click events on `hs.menubar`) be always "immune" from termination; that would seem to require a fourth `protectedCallbackAlwaysAndDisplayError` function that would skip the terminated check.
(One could see things the other way around, and consider the "obey termination" as the exception to the rule; but I'd say there's a lot more of these - at least all the watchers, timer, eventtap...)

**Comment by asmagill** [comment]
I had forgotten about this altogether...

I would be really leery of attempting to "halt" lua callbacks after an error of any type... we have the requests coming from too many different sources... the console; delegates for menus, speech, webviews, etc; objective-c blocks for timers, tasks, etc. from the application dispatch queue; ipc from events (most watchers) and the command line `hs` command; hotkeys; other threads (currently only `hs._asm.serial`, but I've also been experimenting with this to reduce blocking in other areas)... you get the idea.

And while many of these callbacks are primarily for notifications or "side-effects", some expect lua to produce results which then affect the course of application behavior.

To reduce clutter, especially in the console wrt to logs, I would be much more comfortable with adding hooks to the Hammerspoon (i.e. MJLua.m) side of the [skin logDebug:], [skin logWarn:], and [skin logError:] methods that would allow you to add a callback when these occur.  The callback could optionally return a boolean indicating that it was handled and the regular behavior was not required.  These callbacks could do things like adjust logging levels for modules, (if the hs.logger instance per module is visible somewhere), set a flag so that your other callbacks would do something else (not produce output, return immediately, but with a proper "value" where appropriate... of course you would have to modify your other callbacks with this behavior in mind.

BTW (and @cmsj, I'd like your input on this as well), I'm strongly considering adding logInfo and logVerbose (or maybe reduce it all down to one [skin logAtLevel:# withMessage:msg], not sure) to more closely follow hs.logger's levels... then with callbacks, you could even dump the messages into a global "Hammerspoon" instance of hs.logger.

**Comment by cmsj** [comment]
@asmagill +1 for additional logging methods in LuaSkin. I'd suggest we keep log{Info,Verbose,Warn,Error, etc) and have them back onto logAtLevel:withMessage

**Comment by lowne** [comment]
I agree that implementing this thing in each module is too much work and error prone, which is why I argued it should be in LuaSkin (or rather its HS "wrapper"). It could also be done in the Lua side of every "consumer" module (still a lot of work, but fairly straightforward), something like this:
```lua
function hs.callback(fn,...) if not hs._TERMINATED then return fn(...) end end
```
and then in, say, `window/filter.lua`, `local cb=hs.callback someTimer=timer.new(interval,cb(function()...end))` etc. - but clearly this isn't the "natural" approach.

Speaking of `hs.timer.new()`, it's perhaps useful to think of my proposal as generalising the `continueOnError` argument - I'd like to have it, at the very least, on every watcher, but I'd *also* like to have a *global* error state, instead of stopping just the individual timer/watcher when an error occurs within its callback. 
Or in other words: the `continueOnError` argument isn't at all *necessary* - the user could just `pcall` everything within their timer callback, and stop the timer themselves if something goes wrong, but we probably don't want to turn user scripts into Java; I'm merely proposing to extend and generalise this "convenience".

But yes, it's complicated, and everybody got along just fine without it so far, so I don't consider it a must-have. (To address log clutter, after #704 I can just override `hs.showError` and shove `hs.logger.setGlobalLogLevel(0)` in there)

However, on a different topic:
>  some expect lua to produce results which then affect the course of application behaviour

I don't know where these callbacks are, but I'd expect that the relevant module would handle getting back nil as whatever the "do nothing/do the default/just go away" path is for the specific case; if that's not the case, I think it should be; if that's not possible, I'm curious to know what they are.

**Comment by asmagill** [comment]
Off the top of my head, I don't remember which ones.  And yes, most of them (at least the ones I *remember* writing) generally can handle a nil result either as a negative result or at least won't error out on it, but that's a convention,  not a requirement.

And FWIW, I don't want my other code to stop working because one portion fails, so no I wouldn't want it to be in/on by default.  If finding the right log message is hard, then I'm either logging wrong or in the wrong place (you can always output to file or to the system log, though it does take a few extra commands)

And is the rest a long winded way to say that you don't want the log methods to have a callback hook to Hammerspoon?  Or that you do?

**Comment by lowne** [comment]
> Then we'd have an hs.stopOnError variable, **false by default**. [emphasis added for clarity]

Re: log hooks, the way I see it, if the code is calling `log.e`, then obviously it is already aware of the error and should deal with it appropriately by gracefully de-escalating from "exceptional failure state" (that deserves the logging so the user can take appropriate actions if deemed necessary) to "regular failure state" (which the call stack already expects, and knows how to handle); in other words, I don't think modules should call `log.e` and then explode - if an explosion is warranted (usually because of *user* error), they should directly call `error`, and the call stack has always the possibility of catching these via `pcall` and doing something sensible; therefore, with regards to the "terminate on error" topic, I don't have any immediate use for such hooks as they don't really address the issue. But I certainly don't object to them, might be useful down the line.

**Comment by cmsj** [comment]
We definitely have a lot of error logging cleanup to be done, but I don't think we need an issue open to remind us, I'm trying to reduce the number of open issues to things that are actually directly actionable, or enhancements.
